pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,273 @@
+			tenancyv1alpha1.WorkspaceInitialized,
+			tenancyv1alpha1.WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of any APIBindings for an APIExport that may have conflicts. We don't really expect this to
+	// happen, because we're tracking what we've created, but "just in case"...
+	bindingsWithConflicts := make(map[tenancyv1alpha1.APIExportReference]sets.String)
+
+	// This keeps track of which APIBindings have been created for which APIExports
+	exportToBinding := map[apisv1alpha1.WorkspaceExportReference]*apisv1alpha1.APIBinding{}
+
+	// Check for any naming conflicts
+	for i := range bindings {
+		binding := bindings[i]
+
+		if binding.Spec.Reference.Workspace == nil {
+			continue
+		}
+
+		// Track what we have ("actual")
+		exportToBinding[*binding.Spec.Reference.Workspace] = binding
+
+		// Record any conflicts
+		for _, cond := range binding.Status.Conditions {
+			if cond.Reason == apisv1alpha1.NamingConflictsReason {
+				logger.V(2).Info("found an APIBinding with conflicts", "message", cond.Message)
+
+				exportRef := tenancyv1alpha1.APIExportReference{
+					WorkspacePath: binding.Spec.Reference.Workspace.Path,
+					Name:          binding.Spec.Reference.Workspace.ExportName,
+				}
+
+				conflicts, ok := bindingsWithConflicts[exportRef]
+				if !ok {
+					conflicts = sets.NewString()
+					bindingsWithConflicts[exportRef] = conflicts
+				}
+
+				conflicts.Insert(binding.Name)
+
+				continue
+			}
+		}
+	}
+
+	requiredExportRefs := map[tenancyv1alpha1.APIExportReference]struct{}{}
+
+	// Get what we've already created for this workspace (from previous reconcile iteration(s) ). Note that even though
+	// this returns a pointer to the set and not a copy, we are safe to mutate the contents outside the lock because
+	// only 1 worker can ever be working on this workspace at a time.
+	createdBindings := func() map[tenancyv1alpha1.APIExportReference]struct{} {
+		b.createdBindingsLock.Lock()
+		defer b.createdBindingsLock.Unlock()
+
+		createdBindings := b.createdBindings[clusterName]
+		if createdBindings == nil {
+			createdBindings = make(map[tenancyv1alpha1.APIExportReference]struct{})
+			b.createdBindings[clusterName] = createdBindings
+		}
+
+		return createdBindings
+	}()
+
+	for _, cwt := range cwts {
+		logger := logging.WithObject(logger, cwt)
+		logger.V(2).Info("attempting to initialize APIBindings")
+
+		for i := range cwt.Spec.DefaultAPIBindings {
+			exportRef := cwt.Spec.DefaultAPIBindings[i]
+
+			// Keep track of unique set of expected exports across all CWTs
+			requiredExportRefs[exportRef] = struct{}{}
+
+			logger := logger.WithValues("apiExportPath", exportRef.WorkspacePath, "apiExportName", exportRef.Name)
```

Comment:
above (and otherwise in the code) we use e.g. `clusterWorkspaceType.path` - might be nice to do `apiExport.Path` and `apiExport.Name`

Reply:
will do